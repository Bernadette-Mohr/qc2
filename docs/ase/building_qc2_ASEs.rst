.. _build_ASEs:

Building your own qc2-ASE calculator
====================================
