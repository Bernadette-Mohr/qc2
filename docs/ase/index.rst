ASE calculators
===============


.. toctree::
  :maxdepth: 1
