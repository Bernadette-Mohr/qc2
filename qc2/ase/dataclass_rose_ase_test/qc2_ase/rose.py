"""This module defines an ASE interface to Rose.

Original paper & official release:
https://pubs.acs.org/doi/10.1021/acs.jctc.0c00964

GitLab page:
https://gitlab.com/quantum_rose/rose

Note: see also https://pubs.acs.org/doi/10.1021/ct400687b.
"""
from dataclasses import dataclass, field
from enum import Enum
from ase import Atoms
from ase.calculators.calculator import FileIOCalculator
from typing import Optional, List, Union, Sequence


class RoseCalcType(Enum):
    """Enumerator class defining the type of calculation done by Rose."""
    ATOM_FRAG = 'atom_frag'  # calc IAOs from atom-free AOs.
    MOL_FRAG = 'mol_frag'    # calc IFOs from free molecular fragment MOs.


class RoseInterfaceMOCalculators(Enum):
    """Enumerator class defining the program to which Rose is interfaced."""
    PYSCF = 'pyscf'
    PSI4 = 'psi4'
    DIRAC = 'dirac'
    ADF = 'adf'
    GAUSSIAN = 'gaussian'


@dataclass
class RoseInputDataClass:
    """A dataclass representing input options for Rose."""
    rose_target: Atoms
    rose_frags: Union[Atoms, Sequence[Atoms]]

    calculate_mo: bool = True
    rose_calc_type: RoseCalcType = RoseCalcType.ATOM_FRAG.value

    run_postscf: bool = False
    restricted: bool = True
    openshell: bool = False
    relativistic: bool = False

    # Rose intrinsic options => expected to be fixed at the default values?
    version: str = ['Stndrd_2013', 'Simple_2013', 'Simple_2014'][0]
    exponent: int = [2, 3, 4][2]
    spherical: bool = False
    uncontract: bool = True
    test: bool = True
    wf_restart: bool = True
    get_oeint: bool = True
    save: bool = True
    avas_frag: Optional[List[int]] = field(default_factory=list)
    nmo_avas: Optional[List[int]] = field(default_factory=list)

    def __post_init__(self):
        """This method is called after the instance is initialized."""
        if self.test:
            self.get_oeint = True

    """Description of the attributes to be added below."""


class Rose(RoseInputDataClass, FileIOCalculator):
    """A general ASE calculator for ROSE (Reduction of Orbital Space Extent).

    Args:
        RoseInputDataClass (RoseInputDataClass): Tailor-made base dataclass
            for Rose; see above.
        FileIOCalculator (FileIOCalculator): Base class for ASE calculators
            that write/read input/output files.

    """
    implemented_properties: List[str] = []
    command: str = 'echo "Executing Rose...done"'
    restart: bool = None
    ignore_bad_restart_file: bool = FileIOCalculator._deprecated
    label: str = 'rose'
    atoms: Atoms = None

    def __init__(self, *args, **kwargs) -> None:
        """ASE-Rose Class Constructor to initialize the object.

        Give an example here on how to use....
        """
        # initializing base classes
        RoseInputDataClass.__init__(self, *args, **kwargs)
        FileIOCalculator.__init__(self, *args, **kwargs)

    def calculate(self) -> None:
        """Executes Rose workflow."""
        self.get_input_genibo_avas()
        self.get_input_xyz()

        if self.calculate_mo:
            self.get_mo_files()

        self.run_rose()

        if self.save:
            self.save_ibos()

        if len(self.avas_frag) != 0:
            self.run_avas()

        if self.run_postscf:
            self.run_post_hf()

    def get_input_genibo_avas(self) -> None:
        """Generates INPUT_GENIBO & INPUT_AVAS fortran files for Rose."""
        write_input_genibo_avas(self)

    def get_input_xyz(self) -> None:
        """Generates Molecule and Fragment xyz files for Rose."""
        print("Creating Molecule and Frags inputs....done")

    def get_mo_files(self) -> None:
        """Generates atomic and molecular orbitals files for Rose."""
        print('Calculating MO files....done')

    def run_rose(self) -> None:
        """Runs Rose executable 'genibo.x'."""
        print('Executing Rose....done')

    def save_ibos(self) -> None:
        """Generates a checkpoint file ('ibo.chk') with the final IBOs."""
        print('Saving ibos....done')

    def run_avas(self) -> None:
        """Runs 'avas.x' executable."""
        print('Executing avas....done')

    def run_post_hf(self) -> None:
        """Performs CASCI and/or CASSCF calculations."""
        print('CASSCF?....done')


def write_input_genibo_avas(input_data: RoseInputDataClass) -> None:
    """Summary.

    Args:
        input_data (RoseInputDataClass): description
    """
    print("Creating genibo and avas input....done")

    molecule_charge = input_data.rose_target.calc.parameters.charge
    molecule_mo_calculator = input_data.rose_target.calc.name.lower()
    molecule_natom = len(input_data.rose_target.symbols)

    fragments_mo_calculator = []
    for n in range(len(input_data.rose_frags)):
        frag_mo_program = input_data.rose_frags[n].calc.name.lower()
        fragments_mo_calculator.append(frag_mo_program)

    version = input_data.version
    exponent = input_data.exponent
    restricted = input_data.restricted
    test = input_data.test

    avas_frag = input_data.avas_frag
    nmo_avas = input_data.nmo_avas

    # creating INPUT_GENIBO file
    with open("INPUT_GENIBO", "w") as f:
        f.write("**ROSE\n")
        f.write(".VERSION\n")
        f.write(version+"\n")
        f.write(".CHARGE\n")
        f.write(str(molecule_charge)+"\n")
        f.write(".EXPONENT\n")
        f.write(str(exponent)+"\n")
        if not restricted:
            f.write(".UNRESTRICTED\n")
        f.write(".FILE_FORMAT\n")
        f.write(molecule_mo_calculator+"\n")
        if test == 1:
            f.write(".TEST  \n")
        f.write(".AVAS  \n")
        f.write(str(len(avas_frag))+"\n")
        f.writelines("{:3d}".format(item) for item in avas_frag)
        f.write("\n*END OF INPUT\n")

    # creating INPUT_AVAS file
    with open("INPUT_AVAS", "w") as f:
        f.write(str(molecule_natom) + " # natoms\n")
        f.write(str(molecule_charge) + " # charge\n")
        if restricted:
            f.write("1 # restricted\n")
        else:
            f.write("0 # restricted\n")
        f.write("1 # spatial orbs\n")
        f.write(molecule_mo_calculator + " # MO file for the full molecule\n")
        f.write(fragments_mo_calculator + " # MO file for the fragments\n")
        f.write(str(len(nmo_avas)) + " # number of valence MOs in B2\n")
        f.writelines("{:3d}".format(item) for item in nmo_avas)

    # print(molecule_natom)
