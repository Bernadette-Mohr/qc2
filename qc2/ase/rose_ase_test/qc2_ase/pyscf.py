"""This module defines an ASE interface to PySCF.

https://pyscf.org/ => Official documentation
https://github.com/pyscf/pyscf => GitHub page

Note: Adapted from
https://github.com/pyscf/pyscf/blob/master/pyscf/pbc/tools/pyscf_ase.py
&
https://github.com/pyscf/pyscf/issues/624
"""
from typing import Union, Optional, List, Dict, Any
import warnings
import numpy as np
from ase import Atoms
from ase.calculators.calculator import Calculator, all_changes
from ase.units import Ha, Bohr
from ase.calculators.calculator import InputError
from ase.calculators.calculator import CalculatorSetupError


def ase_atoms_to_pyscf(ase_atoms: Atoms) -> List[List[Union[str, np.ndarray]]]:
    """Converts ASE atoms to PySCF atom.

    Args:
        ase_atoms (Atoms): ASE Atoms object.

    Returns:
        List[List[Union[str, np.ndarray]]]: PySCF atom.
    """
    return [[ase_atoms.get_chemical_symbols()[i], ase_atoms.get_positions()[i]]
            for i in range(len(ase_atoms.get_positions()))]


class PySCF(Calculator):
    """An ASE calculator for the PySCF quantum chemistry package.

    Args:
        Calculator (Calculator): Base-class for all ASE calculators.

    Raises:
        InputError: If attributes other than
            'method', 'xc', 'basis', 'multiplicity',
            'charge', and 'verbose' are input as Calculator.
        CalculatorSetupError: If abinitio methods other than
            'scf.RHF', 'scf.UHF', 'scf.ROHF',
            'dft.RKS', 'dft.UKS', and 'dft.ROKS' are selected.

    Example of a typical ASE-PySCF input:

    >>> from ase import Atoms
    >>> from ase.build import molecule
    >>> from qc2.ase.pyscf import PySCF
    >>>
    >>> molecule = Atoms(...) or molecule = molecule('...')
    >>> molecule.calc = PySCF(method='dft.RKS',
                            xc='b3lyp',
                            basis='6-31g*',
                            charge=0,
                            multiplicity=1,
                            verbose=0)
    >>> energy = molecule.get_potential_energy()
    >>> gradient = molecule.get_forces()

    where (currently)

    method = 'scf.RHF'
           = 'scf.UHF'
           = 'scf.ROHF'
           = 'dft.RKS'
           = 'dft.UKS'
           = 'dft.ROKS'
    """
    implemented_properties: List[str] = ['energy', 'forces']

    default_parameters: Dict[str, Any] = {'method': 'dft.RKS',
                                          'basis': '6-31g*',
                                          'xc': 'b3lyp',
                                          'multiplicity': 1,
                                          'charge': 0,
                                          'verbose': 0}

    def __init__(self,
                 restart: Optional[bool] = None,
                 ignore_bad_restart: Optional[bool] = False,
                 label: Optional[str] = 'PySCF',
                 atoms: Optional[Atoms] = None,
                 command: Optional[str] = None,
                 directory: str = '.',
                 **kwargs) -> None:
        """ASE-PySCF Class Constructor to initialize the object.

        Note: Basic implementation based on the
            class Psi4(Calculator); see, e.g., ase/ase/calculators/psi4.py.

        Args:
            restart (bool, optional): Prefix for restart file.
                May contain a directory. Defaults to None: don't restart.
            ignore_bad_restart (bool, optional): Deprecated and will
                stop working in the future. Defaults to False.
            label (str, optional): Calculator name. Defaults to 'PySCF'.
            atoms (Atoms, optional): Atoms object to which the calculator
                will be attached. When restarting, atoms will get its
                positions and unit-cell updated from file. Defaults to None.
            command (str, optional): Command used to start calculation.
                Defaults to None.
            directory (str, optional): Working directory in which
                to perform calculations. Defaults to '.'.
        """
        # initializing base class Calculator.
        # see ase/ase/calculators/calculator.py.
        super().__init__(restart=restart,
                         ignore_bad_restart=ignore_bad_restart, label=label,
                         atoms=atoms, command=command, directory=directory,
                         **kwargs)
        """Transforms **kwargs into a dictionary with calculation parameters.

        Starting with (attr1=value1, attr2=value2, ...)
            it creates self.parameters['attr1']=value1, and so on.
        """

        # Check self.parameters input keys and values
        self.check_pyscf_attributes()

    def check_pyscf_attributes(self) -> None:
        """Checks for any missing and/or mispelling PySCF input attribute.

        Note: it can also be used to eventually set specific
        environment variables, ios, etc.
        """
        recognized_attributes: List[str] = [
            'ignore_bad_restart', 'command', 'method',
            'xc', 'basis', 'multiplicity', 'charge',
            'verbose', 'kpts', 'nbands', 'smearing'
            ]

        # self.parameters gathers all PYSCF input options in a dictionary.
        # it is defined by class Calculator(BaseCalculator)
        # see ase/ase/calculators/calculator.py
        for key, value in self.parameters.items():
            # check attribute name
            if key not in recognized_attributes:
                raise InputError('Attribute', key,
                                 ' not recognized. Please check input.')

        # dealing with lack of multiplicity and charge info
        if 'multiplicity' not in self.parameters.keys():
            self.parameters['multiplicity'] = 1
            warnings.warn('Multiplicity not provided.'
                          'Assuming default singlet.')

        if 'charge' not in self.parameters.keys():
            self.parameters['charge'] = 0
            warnings.warn('Charge not provided. Assuming default zero.')

        # verbose sets the amount of pyscf printing
        # verbose = 0 prints no info
        if 'verbose' not in self.parameters.keys():
            self.parameters['verbose'] = 0

        # dealing with some ASE specific inputs
        if 'kpts' in self.parameters.keys():
            raise InputError('This ASE-PySCF interface does not yet implement'
                             ' periodic calculations, and thus does not'
                             ' accept k-points as parameters. '
                             'Please remove this argument.')

        if 'nbands' in self.parameters.keys():
            raise InputError('This ASE-PySCF interface '
                             'does not support the keyword "nbands".')

        if 'smearing' in self.parameters.keys():
            raise InputError('Finite temperature DFT is not currently'
                             ' implemented in this PySCF-ASE interface,'
                             ' thus a smearing argument cannot be utilized.'
                             ' Please remove this argument.')

    def calculate(self, atoms: Optional[Atoms] = None,
                  properties: List[str] = ['energy'],
                  system_changes: List[str] = all_changes) -> None:
        """This method is the core responsible for the actual calculation.

        Note: Implementation based on Calculator.calculate() method.
            see also ase/ase/calculators/calculator.py.

        Args:
            atoms (Atoms, optional): Atoms object to which
                the calculator is attached. Defaults to None.
            properties (list[str], optional): List of what properties
                need to be calculated. Defaults to ['energy'].
            system_changes (list[str], optional): List of what has changed
                since last calculation. Can be any combination
                of these six: 'positions', 'numbers', 'cell', 'pbc',
                'initial_charges' and 'initial_magmoms'.
                Defaults to all_changes.

        Raises:
            CalculatorSetupError: If a proper geometry is not provided.
            CalculatorSetupError: If abinitio methods other than
            'scf.RHF', 'scf.UHF', 'scf.ROHF', 'dft.RKS', 'dft.UKS',
                and 'dft.ROKS' are selected.
        """
        implemented_methods: List[str] = ['scf.RHF', 'scf.UHF', 'scf.ROHF',
                                          'dft.RKS', 'dft.UKS', 'dft.ROKS']

        from pyscf import gto  # , scf, dft

        # setting up self.atoms attribute from base class Calculator.calculate.
        # this is extracted from the atoms Atoms object.
        super().calculate(atoms=atoms)

        # checking that self.atoms has been properly initiated/updated
        if self.atoms is None:
            raise CalculatorSetupError('An Atoms object must be provided to '
                                       'perform a calculation.')

        # the spin variable corresponds to 2S instead of 2S+1
        spin_2s = self.parameters['multiplicity'] - 1

        # passing geometry and other definitions
        molecule = gto.M(atom=ase_atoms_to_pyscf(self.atoms),
                         basis=self.parameters['basis'],
                         charge=self.parameters['charge'],
                         spin=spin_2s)

        # checking wf input name => this is case sensitive
        if self.parameters['method'] in implemented_methods:
            self.wf = eval(self.parameters['method'])(molecule)
        else:
            raise CalculatorSetupError('Method not yet implemented. '
                                       'Current PySCF-ASE calculator '
                                       'only allows for',
                                       implemented_methods,
                                       'wave functions.'
                                       ' Please check input method.')

        self.wf.verbose = self.parameters['verbose']

        if 'dft' in self.parameters['method']:
            self.wf.xc = self.parameters['xc']

        # calculating energy in eV
        energy = self.wf.kernel() * Ha
        self.results['energy'] = energy

        # calculating forces
        if 'forces' in properties:
            gf = self.wf.nuc_grad_method()
            gf.verbose = self.parameters['verbose']
            if 'dft' in self.parameters['method']:
                gf.grid_response = True
            # analytic gradienta in eV/Angstrom
            forces = -1.0 * gf.kernel() * (Ha / Bohr)
            totalforces = []
            totalforces.extend(forces)
            totalforces = np.array(totalforces)
            self.results['forces'] = totalforces
