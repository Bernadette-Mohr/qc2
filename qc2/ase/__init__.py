"""qc2 ASE package."""
from .pyscf import PySCF
from .rose import ROSE, ROSETargetMolecule, ROSEFragment
from .dirac import DIRAC
