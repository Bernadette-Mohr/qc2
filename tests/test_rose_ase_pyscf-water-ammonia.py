from ase import Atoms
from qc2.ase.rose import Rose
from qc2.ase.pyscf import PySCF

import os
import subprocess

def clean_stuff():
    command = "rm *xyz *dfcoef DFCOEF* *inp INPUT* \
        MOLECULE.XYZ MRCONEE* *dfpcmo DFPCMO* *fchk \
        *in fort.100 timer.dat INFO_MOL *.pyscf \
        IAO_Fock SAO *.npy *.clean *\ 2* OUTPUT_AVAS \
        *.chk ILMO*dat"
    subprocess.Popen(command, shell=True, cwd='.')

def run_water_amonia_rose():
    """ Water-Ammonia example calculation."""

    # define target molecule
    mol = Atoms('NH3OH2',
                positions=[[-1.39559, -0.02156,  0.00004],
                           [-1.62981,  0.96110, -0.10622],
                           [-1.86277, -0.51254, -0.75597],
                           [-1.83355, -0.33077,  0.86231],
                           [ 1.56850,  0.10589,  0.00001],
                           [ 0.60674, -0.03396, -0.00063],
                           [ 1.94052, -0.78000,  0.00022]
                           ],
                calculator=PySCF(method='scf.RHF',
                                 basis='unc-sto-3g',
                                 multiplicity=1)
                )

    # define molecular fragments
    frag1 = Atoms('NH3',
                  positions=[[-1.39559, -0.02156,  0.00004],
                             [-1.62981,  0.96110, -0.10622],
                             [-1.86277, -0.51254, -0.75597],
                             [-1.83355, -0.33077,  0.86231]
                             ],
                  calculator=PySCF(method='scf.RHF',
                                   basis='unc-sto-3g',
                                   multiplicity=1
                                   )
                  )

    frag2 = Atoms('OH2',
                  positions=[[1.56850,  0.10589,  0.00001],
                             [0.60674, -0.03396, -0.00063],
                             [1.94052, -0.78000,  0.00022]
                             ],
                  calculator=PySCF(method='scf.RHF',
                                   basis='unc-sto-3g',
                                   multiplicity=1
                                   )
                  )

    # perform scf calculations on all species
    mol.get_potential_energy()
    frag1.get_potential_energy()
    frag2.get_potential_energy()

    # prepare Rose-ASE calculator
    rose_calc = Rose(rose_calc_type='mol_frag',
                     rose_target=mol,
                     rose_frags=[frag1, frag2],
                     test = True,
                     avas_frag=[0], nmo_avas=[3, 4, 5],
                     frag_valence=[[1,7],[2,6]],
                     frag_core=[[1,1],[2,1]]
                     )

    # run the calculator
    rose_calc.calculate()

def test_rose_output():

    clean_stuff()

    expected_output = open(
        'test_rose_ase_pyscf-water-ammonia.out').read()

    run_water_amonia_rose()

    output = open('rose.out').read()

    clean_stuff()

    assert output == expected_output
