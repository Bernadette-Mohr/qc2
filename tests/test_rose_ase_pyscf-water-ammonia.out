
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @       Reduction of Orbital Space Extent        @
 @                     (ROSE)                     @
 @                                                @
 @ By Bruno Senjean, Souloke Sen, Michal Repisky, @
 @        Gerald Knizia and Lucas Visscher.       @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

 Select algebra (rcq) based on the use of restricted and spatial orbitals:
   rcq = 1, 2 or 4  : real, complex or quaternion algebra, respectively
   restricted       :  T
   spatial_orbitals :  T
   rcq              :  1

 ####################################
 #      Molecular  Coordinates      #
 ####################################

Atom    Z           X           Y           Z
N    7.00 -2.63728288 -0.04074250  0.00007559
H    1.00 -3.07989454  1.81621578 -0.20072671
H    1.00 -3.52012513 -0.96856023 -1.42857626
H    1.00 -3.46490734 -0.62506471  1.62952973
H    1.00  1.14657243 -0.06417510 -0.00119053
H    1.00  3.66705134 -1.47398638  0.00041574
O    8.00  2.96403543  0.20010310  0.00001890

 ####################################
 #         Basis set size           #
 #               and                #
 #      Canonical MO Energies       #
 ####################################

 No. basis functions    B1:     45
 No. molecular orbitals B1:     43
 No. occupied  orbitals B1:      8
 No. molecular orbitals B2:     13


 ####################################
 #   Intrinsic Fragments Orbitals   #
 ####################################

  Version: Standard 2013

  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.5025E-14


  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.7058E-14

 ####################################
 #  Partial charges of fragments    #
 ####################################

       Fragment      Charge
   frag0           0.144506
   frag1          -0.144506



 ####################################
 # Intrinsic Localized occupied MOs #
 ####################################


 ----- Percentage of localization before the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1     -1.2682 [    2]
                [ 99.8]
  2     -1.1641 [    1    2]
                [ 98.6  1.4]
  3     -0.6368 [    1    2]
                [ 96.7  3.3]
  4     -0.6365 [    1    2]
                [ 97.4  2.6]
  5     -0.6183 [    1    2]
                [  9.5 90.5]
  6     -0.4782 [    1    2]
                [ 16.8 83.2]
  7     -0.4207 [    1    2]
                [ 73.5 26.5]
  8     -0.4041 [    2]
                [100.0]

 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        6.158230    7.17E-02
    2        7.711248    1.33E-02
    3        7.734115    2.43E-03
    4        7.734697    1.55E-04
    5        7.734699    5.12E-10
    6        7.734699    5.62E-16

 ******************************************
 Final converged values (it, func, fGrad):
    6            7.7346993569281421  5.6233E-16
 ******************************************


 ----- Percentage of localization after the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1]
    [100.0]
  2 [    1]
    [100.0]
  3 [    1]
    [100.0]
  4 [    1    2]
    [ 92.7  7.3]
  5 [    2]
    [100.0]
  6 [    2]
    [100.0]
  7 [    2]
    [ 99.9]
  8 [    2]
    [100.0]

 ----- Percentage of localization after recanonization -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1     -1.1637 [    1    2]
                [ 98.9  1.1]
  2     -0.6367 [    1]
                [100.0]
  3     -0.6356 [    1]
                [100.0]
  4     -0.4388 [    1    2]
                [ 93.9  6.1]
  5     -1.2677 [    2]
                [100.0]
  6     -0.6128 [    2]
                [ 99.9]
  7     -0.4675 [    2]
                [100.0]
  8     -0.4041 [    2]
                [100.0]


 ############################################
 # Intrinsic Localized valence virtual MOs  #
 ############################################


 ----- Percentage of localization before the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1      0.2283 [    1    2]
                [  7.9 92.1]
  2      0.3208 [    1    2]
                [ 54.8 45.2]
  3      0.3334 [    1    2]
                [ 85.7 14.3]
  4      0.3357 [    1    2]
                [ 61.4 38.6]
  5      0.4894 [    1    2]
                [ 97.4  2.6]

 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        2.455536    1.44E-01
    2        4.733612    8.88E-03
    3        4.734699    3.66E-06
    4        4.734699    6.22E-13
    5        4.734699    1.62E-16

 ******************************************
 Final converged values (it, func, fGrad):
    5            4.7346993564360957  1.6237E-16
 ******************************************


 ----- Percentage of localization after the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1]
    [100.0]
  2 [    1]
    [ 99.9]
  3 [    1]
    [100.0]
  4 [    1    2]
    [  7.3 92.7]
  5 [    2]
    [100.0]

 ####################################
 #      Additional virtuals         #
 ####################################

 No. molecular virtuals B3:     30
 No. selected  virtuals B3:     12
 No. reference virtuals B4:     25


 ####################################
 #   Intrinsic Fragments Orbitals   #
 ####################################

  Version: Standard 2013

  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.6953E-14


  --- Check Symmetrical Orthogonalization ---
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.1324E-13


 ##############################################
 # Intrinsic Localized additional virtual MOs #
 ##############################################


 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        7.981183    3.04E-02
    2       11.375314    7.75E-03
    3       11.436208    2.11E-03
    4       11.438642    6.09E-04
    5       11.438844    8.66E-07
    6       11.438844    1.30E-09
    7       11.438844    1.22E-14
    8       11.438844    7.59E-18

 ******************************************
 Final converged values (it, func, fGrad):
    8           11.4388443164368496  7.5893E-18
 ******************************************


 ----- Percentage of localization after the localization procedure -----

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1]
    [100.0]
  2 [    1]
    [100.0]
  3 [    1]
    [100.0]
  4 [    1]
    [100.0]
  5 [    1]
    [100.0]
  6 [    1]
    [100.0]
  7 [    1    2]
    [ 81.8 18.2]
  8 [    2]
    [100.0]
  9 [    2]
    [100.0]
 10 [    2]
    [100.0]
 11 [    2]
    [100.0]
 12 [    2]
    [ 99.8]

 ----- Recanonicalized virtual energies of fragments -----

    Fragment          MO              Energy
           1           1         0.238657888
           1           2         0.333404572
           1           3         0.335695069
           1           4         0.795796029
           1           5         0.797974507
           1           6         1.012927063
           1           7         1.318069521
           1           8         1.353480374
           1           9         1.357502755
           1          10         1.751989751
           2           1         0.317395753
           2           2         0.485932207
           2           3         0.939210358
           2           4         1.027959054
           2           5         1.591697181
           2           6         1.646503923
           2           7         1.836937715

 ####################################
 #         Starting TESTS           #
 ####################################

 The matrix is normalized.
 The matrix is normalized.
 The matrix is normalized.
 Deviation of IBO_RDM from HF_RDM  :   1.9885549835285585E-014
 Deviation from orthogonality      :   8.2430099525239453E-015
 Deviation from orthogonality      :   8.9116984521814421E-015
 Deviation in number of electrons  :  -3.5527136788005009E-015
 Deviation in number of electrons  :  -1.4210854715202004E-014
 HF energy (moB1)   :   -130.74589367992581
 HF energy (IAO)    :   -130.74589367992576
 HF energy (IBO/MO) :   -130.74589367992573
 Deviation from the HF energy in the original basis and in the IAO basis:   5.68434189E-14
 Deviation from the HF energy in the original basis and in the IBO basis:   8.52651283E-14

 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @       Reduction of Orbital Space Extent        @
 @                                                @
 @                      END                       @
 @                                                @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @        Automated Valence Active Space          @
 @                    (AVAS)                      @
 @                                                @
 @               By Bruno Senjean                 @
 @                                                @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

 Threshold:   0.10E-13
 Select algebra (rcq) based on the use of restricted and spatial orbitals:
   rcq = 1, 2 or 4  : real, complex or quaternion algebra, respectively
   restricted       :  T
   spatial_orbitals :  T
   rcq              :  1

 ####################################
 #         Basis set size           #
 ####################################

 No. basis functions    B1:     45
 No. molecular orbitals B1:     45
 No. occupied  orbitals B1:     10
 No. basis functions    B2:     45
 No. molecular orbitals B2:      3


 ####################################
 #    Active space construction     #
 ####################################

 No. of frozen occupied:      7
 No. of active occupied:      3
 No. of active virtuals:      3
 No. of frozen virtuals:     32

 --------- Eigenvalues of the Projected overlap matrices (active MOs) ---------

  MO                   eigval
   8       0.0008945760010692
   9       0.9374975401260698
  10       0.9998509006652673
  11       0.9991054239989300
  12       0.0625024598739295
  13       0.0001490993347300

 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @        Automated Valence Active Space          @
 @                    (AVAS)                      @
 @                                                @
 @                     END                        @
 @                                                @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
output file: /Users/murilo/Library/CloudStorage/OneDrive-NetherlandseScienceCenter/Documents/NLeSC_Projects/qc2nl_quantum_computing_for_quantum_chemistry/rose/examples/PYSCF/H5-N1-O1_0.out
Hartree-Fock energy: -130.7458927644741
output file: /Users/murilo/Library/CloudStorage/OneDrive-NetherlandseScienceCenter/Documents/NLeSC_Projects/qc2nl_quantum_computing_for_quantum_chemistry/rose/examples/PYSCF/H3-N1_0.out
Hartree-Fock energy: -55.56739074987616
output file: /Users/murilo/Library/CloudStorage/OneDrive-NetherlandseScienceCenter/Documents/NLeSC_Projects/qc2nl_quantum_computing_for_quantum_chemistry/rose/examples/PYSCF/H2-O1_0.out
Hartree-Fock energy: -75.16209450502814
overwrite output file: /Users/murilo/Library/CloudStorage/OneDrive-NetherlandseScienceCenter/Documents/NLeSC_Projects/qc2nl_quantum_computing_for_quantum_chemistry/rose/examples/PYSCF/H5-N1-O1_0.out
Hartree-Fock energy: -130.74589276447398

--------------------------------------------------------------------------
------------------------ POST-SCF CALCULATIONS ---------------------------
--------------------------------------------------------------------------

CASCI (original): -130.7471862792149
CASCI (ibo     ): -130.74589278716073
CASCI (avas    ): -130.75696519430144
